Adding pipeline variables metrics & refactored general logic implementation

inge4pres opened the pull request:
Hello,

here a draft of adding pipeline variables inside metrics.
The PR contains a pretty big refactoring for 2 areas:
* initial project polling
* metrics registration

One thing that I'd like to tackle is how to simplify the gauges increment when polling the projects: right now for my understanding of the codebase we are not able to discern if a counter for a project should be incremented every time we process it, because we should fetch the previous value?
You can see this in `emitPipelineVariablesMetric` function: when the first time it runs, it will set a value for the parsed jobs with variables. How could it know the second time it runs that a previous metric has been already set? Does prometheus lib handle that?

mvisonneau commented:
Thanks for these really cool improvements @inge4pres ! Indeed I ran into that problem in the past too, I am not sure there is a way to fetch that value straight from the prometheus library. Perhaps we will have to maintain a copy of that state :(

I also figured that there is currently a bug on the `pollProjectRefsFromPipelines()` function trigger which happens all the time instead of only at the init https://github.com/mvisonneau/gitlab-ci-pipelines-exporter/blob/master/cmd/gitlab.go#L227

Having a better approach would also help fixing this issue as well.

inge4pres commented:
Yes I intentionally introduced the polling at every run: I saw some duplication and thought that it was a good idea to only fetch the refs from the pipeline and pass all of them to the polling.
I know that this is running at every polling interval and would probably move that part higher in the call stack, will push a commit to show what I mean.

inge4pres commented:
Hey there 😄 last commit I sent on this PR is quite huge but I find it very valuable: I fixed the problem of generating the metrics from latest pipelines on init.
I also removed some duplication around fetching the pipelines from the API, extracting a method. It looks like the draft status can be solved now, and also the pipeline variables are now part of the metrics.

mvisonneau commented on cmd/exporter.go:
```diff
@@ -191,25 +71,28 @@ func Run(ctx *cli.Context) error {
 	// Configure liveness and readiness probes
 	health := healthcheck.NewHandler()
-	if !cfg.Gitlab.SkipTLSVerify {
+	if !cfg.Gitlab.SkipHealthCheck {
```
I think it would be nice to have a switch with more precise log.Warn here instead of solely replacing the statement on `cfg.Gitlab.SkipTLSVerify`

inge4pres replied:
Ah the message was not updated... thanks 👍 
Changed to ```
"GitLab health check has been disabled. Readiness checks won't be operated."```

mvisonneau replied:
thanks! we are still missing the case `cfg.Gitlab.SkipTLSVerify == true` though :)

mvisonneau commented on cmd/metrics.go:
```diff
@@ -0,0 +1,173 @@
+	}
+}
+
+func emitStatusMetric(metric *prometheus.GaugeVec, labelValuess []string, statuses []string, status string, sparseMetrics bool) {
```
is `labelValuess` a misspell or on purpose ? 🤔

inge4pres replied:
nope, thanks 👍

mvisonneau commented on cmd/gitlab.go:
```diff
@@ -9,303 +9,77 @@ import (
-	}
-	return false
-}
+var statusesList = []string{"running", "pending", "success", "failed", "canceled", "skipped", "manual"}
```
```suggestion
var statusesList = [...]string{"running", "pending", "success", "failed", "canceled", "skipped", "manual"}
```

as this slice will not expand further

inge4pres replied:
👍

inge4pres replied:
this actually breaks the function using it... I'd keep the slice that is more flexible

mvisonneau replied:
ok 👍

mvisonneau commented on cmd/gitlab.go:
```diff
@@ -9,303 +9,77 @@ import (
 }
 
-func (c *Client) pollPipelineJobs(p Project, pipelineID int, topics string, ref string) error {
+func (c *Client) pollPipelineJobs(gp *gitlab.Project, pipelineID int, topics string, ref string, outputSparseMetricsStatus bool) error {
```
I see that you are using `gitlabProject` in another function: https://github.com/mvisonneau/gitlab-ci-pipelines-exporter/pull/93/files#diff-fb2ba0b24813a3ba03c719c857bd28d1R141

I do not mind the terminology but I reckon it would be nice to have a single way to name this object across the codebase

inge4pres replied:
Agreed! renamed to `gp` to keep it consistent 👍

mvisonneau commented on cmd/gitlab.go:
```diff
@@ -362,74 +138,84 @@ func (c *Client) pollPipelineJobs(p Project, pipelineID int, topics string, ref
 				log.Info("stopping projects polling...")
 				return
+			case todo := <-getRefsOnInit:
+				log.Debugf("executing init pipeline? %v", todo)
```
I believe this should be within the if statement underneath. Also, a more pertinent message could be nice!

inge4pres replied:
agree on the message but the reason for it to be external is to debug what value is passed _before_ evaluating.

Corrected with
```
log.Debugf("should we poll the more recent refs from the last executed pipelines? %v", todo)
```

mvisonneau replied:
oh my bad 🤦 sounds good 👍

mvisonneau commented on cmd/gitlab.go:
```diff
@@ -438,44 +224,59 @@ func (c *Client) pollProjects(until <-chan bool) {
+			log.Errorf("could not get GitLab project with name %s: %v", p.Name, err)
+			continue
+		}
+		pipeRefs, err := c.refsFromPipelines(gitlabProject, cfg.OnInitFetchRefsFromPipelinesDepthLimit)
```
it caused me a braincycle to understand that pipe meant pipeline, perhaps we should use the full word ? 🤷

inge4pres replied:
legit 👍

inge4pres commented on .gitignore:
```diff
@@ -5,4 +5,5 @@ vendor
 coverage.out
-.test
\ No newline at end of file
+testconfig.yml
```
ahhh good I missed it, removed

inge4pres commented on CHANGELOG.md:
```diff
@@ -11,6 +11,7 @@ and this project adheres to [0ver](https://0ver.org).
 
 - Configuration for OpenMetrics Encoding in metrics HTTP endpoint. Enabled by default but can be disable using `disable_openmetrics_encoding: true`.
 - Worker pool for projects polling: set `maximum_projects_poller_workers` with an integer value to control parallelism (defaults to `runtime.GOMAXPROCS(0)`)  
+- Augmented `skip_tls_verify` with `skip_health_check` additional parameter to drive the behaviour of checking healthiness of target service 
```
well initially, the same bool was used for adding the `TransportConfig` option to the http client _and_ starting the health check.
I thought they are separate concepts:
- skip_tls_verify: removes the TLS verification in health_check_url
- skip_health_check: do not do health check at all

So even if you want to use health check, but have a wrong TLS setup (slef-signed cert is the most common case I can think of) you can still perform health checks on the target service.

mvisonneau replied:
yes I realized that it could be confusing. However `skip_tls_verify` main purpose is not to disable the TLS verification in the health check but in the actual API calls. I abused it a bit as the library used for the healthcheck did not support (I am not sure if it is still the case) disabling TLS verification.

inge4pres replied:
I'd keep it as it is if it's ok for you

mvisonneau replied:
yeah no worries 👍

inge4pres commented on README.md:
```diff
@@ -33,7 +33,14 @@ gitlab:
-  # disable TLS verification
+  # disable verification of readiness for target GitLab instance calling `health_url`
+  # skip_health_check: false
+<<<<<<< HEAD
```
ouch :)
thanks, fixed

inge4pres commented on cmd/config_test.go:
```diff
@@ -52,7 +51,7 @@ gitlab:
   token: xrN14n9-ywvAFxxxxxx
   health_url: https://gitlab.example.com/-/health
-  skip_tls_verify: true
+  skip_health_check: true
```
intended: it's tested at line 96, then the other config is tested in the following test `TestParseDefaultsValues`

inge4pres replied:
if you want to refine the testing of these 2 we can introduce another test

mvisonneau replied:
if we introduce a switch case for determining if healthcheck should run or not it could be nice to have a test for it I guess

mvisonneau replied:
just getting it, indeed let's leave it as is 👍 🤦

mvisonneau commented on cmd/exporter.go:
```diff
@@ -191,25 +71,28 @@ func Run(ctx *cli.Context) error {
 	// Configure liveness and readiness probes
 	health := healthcheck.NewHandler()
-	if !cfg.Gitlab.SkipTLSVerify {
+	if !cfg.Gitlab.DisableHealthCheck {
```
in the scenario where `DisableHealthCheck == false`, `DisableTLSVerify == true` and the `GitLab API does not have a valid certificate`, it will prevent it from succeeding the check

inge4pres replied:
does it? I was not able to test it... I might add one

mvisonneau replied:
yes because `healthcheck.HTTPGetCheck(cfg.Gitlab.HealthURL, 5*time.Second)` will not validate it.

inge4pres replied:
got your point 👍 thanks, will fix it implementing `Check` in a func that will use the same transport as GitLab client

inge4pres replied:
this should be ok now

mvisonneau replied:
really cool, thanks for the improvement!

mvisonneau commented on cmd/metrics.go:
```diff
@@ -0,0 +1,173 @@
+
+type pipelineVarsFetchOp func(interface{}, int, ...gitlab.RequestOptionFunc) ([]*gitlab.PipelineVariable, *gitlab.Response, error)
+
+func emitPipelineVariablesMetric(c *Client, metric *prometheus.GaugeVec, projectName, ref string, projectID int, pipelineID int, fetch pipelineVarsFetchOp) error {
```
This could lead to a very large amount of data being fetched. I believe that most of the time, the user will not need it or only a few of them. It could be nice to add a regex to filter-in the CI variable that the user want to get on a per project/wildcard basis. What do you think?

inge4pres replied:
ok, can be an improvement.
would you like to gate in a config the read of such metrics?

mvisonneau replied:
yes that would be great! 🙏 🙇

inge4pres replied:
awesome, whilst at it I was thinking that actually instead of duplicating `gitlab_ci_pipeline_run_count_with_variable` since `gitlab_ci_pipeline_run_count` already exist we could add the labels there (provided config flag is given and filtered with regex)!
WDYT?

inge4pres replied:
implemented the gating and filtering, probably we can keep separated for now

mvisonneau replied:
I agree let's add it straight into `gitlab_ci_pipeline_run_count` 👍

inge4pres replied:
I'll see what's the best way of doing it

mvisonneau commented on CHANGELOG.md:
```diff
@@ -11,6 +11,7 @@ and this project adheres to [0ver](https://0ver.org).
 - Worker pool for projects polling: set `maximum_projects_poller_workers` with an integer value to control parallelism (defaults to `runtime.GOMAXPROCS(0)`)  
+- Augmented `disable_tls_verify` with `disable_health_check` additional parameter to drive the behaviour of checking healthiness of target service 
 - Configurable ServiceMonitor resource through the helm chart
 
```
We should add a line to mention the polling of the CI variables as well.

inge4pres replied:
will add the proper text with regards to what we decide to do with gating

inge4pres replied:
updated, let me know what you think

mvisonneau replied:
awesome thanks!

inge4pres commented:
Last commit resolved a bug introduced in this PR with the following stacktrace:

```
goroutine 397215 [running]:
github.com/mvisonneau/gitlab-ci-pipelines-exporter/cmd.(*Client).pollProjectRefOn(0xc00020b280, 0xc0005b8480, 0xc00051b1c8, 0x6, 0xc001540101, 0x2, 0x2)
	/build/cmd/gitlab.go:157 +0x1145
github.com/mvisonneau/gitlab-ci-pipelines-exporter/cmd.(*Client).pollProject(0xc00020b280, 0xc0000de7b0, 0x25, 0xc000220498, 0x2, 0x0, 0x0, 0xc0005b8480, 0x1, 0x101)
	/build/cmd/gitlab.go:71 +0x36b
github.com/mvisonneau/gitlab-ci-pipelines-exporter/cmd.(*Client).pollProjectsWith.func1(0xc000984d80, 0xc000290360, 0xc000d1c180, 0xc000db1840, 0xc00057eba0)
	/build/cmd/gitlab.go:288 +0x153
created by github.com/mvisonneau/gitlab-ci-pipelines-exporter/cmd.(*Client).pollProjectsWith
	/build/cmd/gitlab.go:282 +0x102
```

mvisonneau commented:
I had a closer look in the use case. I am not sure that having the pipelines variables only on a single metric actually addresses the issue that some users may be facing. It will not be possible to correllate within Prometheus which metric aligns with which if they do not all have the variables within them. 🤔

inge4pres commented:
You mean that it would be difficult for example to extract the % of pipelines with a certain variable compared to the total? that can be done with a `group_left` after a variable filter

inge4pres commented:
Did the weekend bring advices?

mvisonneau commented:
I was more looking into the use case of being able to filter out the status of a pipeline with a particular `ref + variable` combination. This would not be possible using a relational query.

I am happy to see that done in another PR though as this one is getting quite big and covers some other topics as well.

inge4pres commented:
so the status is not yet part of that metric but I guess we can tackle unifying those 2 metrics in a separate PR with a dedicated issue

inge4pres commented:
great job in master! I see you've re-organized the code into packages, awesome stuff!
